blog/_posts/2019-11-13-local-resource.md

```diff
@@ -0,0 +1,146 @@
+
+```python
+local_resource('yarn', cmd='yarn install', deps=['package.json'])
+```
+
+See also: [`local_resource` API spec](api.html#api.local_resource).
+
+## Run commands locally on file change
+Most uses of Tilt focus on easily and quickly seeing your code running in the cloud.
+For some kinds of development, however, you might want to harness
+Tilt's responsiveness to file changes to run commands _locally_ instead. Here are two
+two general cases where running commands locally and responsively might come in handy:
+- *file changes or other artifacts that you want on your local machine (e.g. for
+git commit)*. For instance: when `package.json` changes, run `yarn install`
+(because you want the resulting `yarn.lock` file to exist on your local computer
+so that you can commit it to git).
+
+```python
+local_resource('yarn', cmd='yarn install', deps=['package.json'])
+```
+- *tooling that you have locally but don't want to put on your container images*.
+For instance: rather than pulling/pushing big container images with the Go compiler,
+you'd rather compile your binary locally, and pull that compiled binary into your Docker image
+(or directly into your running container via Live Update).
+
+```python
+local_resource('compile-binary',
+    cmd='go build -i -o ./bin/the-binary github.com/my-org/my-app/',
+    deps='./my-app'
+)
+docker_build('gcr.io/my-org/my-app', context='./bin')
+```
+
+(Of course, the command you run doesn't have to be on that (only) affects your
+local filesystem. It might, for instance, be a script that is _invoked locally_
+but runs against your k8s cluster.)
+
+### Local Resource + Live Update = <3
+
+Local Resource and Live Update are two great tastes that taste great together.
+A useful pattern is to have a local resource which creates an artifact
+that is then picked up by a Live Update and synced to a running container.
+For example, here's how you might use this combination to generate protobufs:
+```python
+# whenever `helloworld.proto` changes, generate Python protobufs
+local_resource('proto',
+    cmd='protoc --proto_path=src --python_out=build/gen',
+    deps=['helloworld.proto']
+)
+
+# define your hello-world resource: k8s yaml + a Docker build
+k8s_yaml('hello-world.yaml')
+docker_build('hello-world', '.',
+    ignore='helloworld.proto',  # don't step on the local resource's toes
+    live_update=[ sync('.', '/app') ]
+)
+```
+
+Note that the `docker_build` call specifies `ignore='helloworld.proto'`. This is
+because we DON'T want an edit to that proto file to _directly_ kick off an update to
+our Docker image (in this case, a Live Update). Rather, an edit to `helloworld.proto`
+triggers the local resource to generate protobufs; when these protobufs appear on disk,
```

Comment:
"protobufs" is a pretty ambiguous term and "when these protobufs appear on disk" might be a bit weird here - outside the context in this sentence, does "protobufs" mean the .proto file? the .pb.go? an instance of a type defined in a .pb.go file? Really, any of the above. In context, one can figure it out, but for someone not very familiar with protobufs or the term's ambiguity, it leaves potential confusing interpretation of this "when helloworld.proto appears on disk, it registers as a file change".

We could instead say "when the protoc-generated code appears on disk" and not worry about it.